File: src/main/java/org/trp/shincolle/utility/ShipTeleportHelper.java
package org.trp.shincolle.utility;

import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.Mth;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;

public final class ShipTeleportHelper {

    private static final double MIN_PLAYER_DISTANCE_SQ = 9.0D;
    private static final double TARGET_REACH_DISTANCE_SQ = 144.0D;
    private static final int[][] PREFERRED_OFFSETS = {
            {-4, 0}, {-4, -2}, {-4, 2},
            {-5, 0}, {-5, -3}, {-5, 3},
            {-3, -3}, {-3, 3}, {-6, -1}, {-6, 1},
            {-2, -4}, {-2, 4}
    };
    private static final int[][] FALLBACK_OFFSETS = {
            {-7, 0}, {-7, -2}, {-7, 2},
            {-5, -5}, {-5, 5}, {-1, -5}, {-1, 5},
            {0, -6}, {0, 6}, {-8, 0}
    };
    private static final int[][] POINT_OFFSETS = {
            {0, 0}, {1, 0}, {-1, 0}, {0, 1}, {0, -1},
            {2, 0}, {-2, 0}, {0, 2}, {0, -2},
            {2, 1}, {2, -1}, {-2, 1}, {-2, -1},
            {1, 2}, {-1, 2}, {1, -2}, {-1, -2},
            {3, 0}, {-3, 0}, {0, 3}, {0, -3}
    };

    private ShipTeleportHelper() {
    }

    public static boolean teleportNearLiving(Entity entity, LivingEntity anchor, double verticalOffset) {
        if (!(entity.level() instanceof ServerLevel serverLevel)) {
            return false;
        }

        Vec3 basePos = new Vec3(anchor.getX(), anchor.getY() + verticalOffset, anchor.getZ());
        Vec3 facing = anchor.getLookAngle();
        Vec3 horizontalFacing = new Vec3(facing.x, 0.0D, facing.z);
        if (horizontalFacing.lengthSqr() < 1.0E-4D) {
            horizontalFacing = new Vec3(0.0D, 0.0D, 1.0D);
        } else {
            horizontalFacing = horizontalFacing.normalize();
        }
        Vec3 right = new Vec3(-horizontalFacing.z, 0.0D, horizontalFacing.x);

        Vec3 candidate = findCandidate(serverLevel, entity, anchor, basePos, horizontalFacing, right, PREFERRED_OFFSETS, true);
        if (candidate == null) {
            candidate = findCandidate(serverLevel, entity, anchor, basePos, horizontalFacing, right, FALLBACK_OFFSETS, true);
        }
        if (candidate == null) {
            candidate = findVerticalFallback(serverLevel, entity, anchor, basePos, horizontalFacing);
        }
        if (candidate == null) {
            return false;
        }

        entity.teleportTo(candidate.x, candidate.y, candidate.z);
        return true;
    }

    public static boolean teleportNearPoint(Entity entity, Vec3 anchor, double verticalOffset) {
        if (!(entity.level() instanceof ServerLevel serverLevel)) {
            return false;
        }

        Vec3 candidate = findPointCandidate(serverLevel, entity, anchor.add(0.0D, verticalOffset, 0.0D));
        if (candidate == null) {
            return false;
        }

        entity.teleportTo(candidate.x, candidate.y, candidate.z);
        return true;
    }

    private static Vec3 findCandidate(ServerLevel level, Entity entity, LivingEntity anchor, Vec3 basePos, Vec3 back, Vec3 right, int[][] offsets, boolean rejectFront) {
        for (int[] offset : offsets) {
            Vec3 candidate = basePos
                    .add(back.scale(offset[0]))
                    .add(right.scale(offset[1]));
            Vec3 safePos = validateCandidate(level, entity, anchor, candidate, back, rejectFront);
            if (safePos != null) {
                return safePos;
            }
        }
        return null;
    }

    private static Vec3 validateCandidate(ServerLevel level, Entity entity, LivingEntity anchor, Vec3 candidate, Vec3 facing, boolean rejectFront) {
        double dx = candidate.x - anchor.getX();
        double dz = candidate.z - anchor.getZ();
        double horizontalDistSq = dx * dx + dz * dz;
        if (horizontalDistSq < MIN_PLAYER_DISTANCE_SQ) {
            return null;
        }

        if (rejectFront) {
            double dot = dx * facing.x + dz * facing.z;
            if (dot > 0.0D) {
                return null;
            }
        }

        if (anchor.distanceToSqr(candidate) > TARGET_REACH_DISTANCE_SQ) {
            return null;
        }

        BlockPos baseBlock = BlockPos.containing(candidate.x, candidate.y, candidate.z);
        for (int dy = 2; dy >= -2; dy--) {
            BlockPos testPos = baseBlock.offset(0, dy, 0);
            if (!canStandAt(level, entity, testPos)) {
                continue;
            }
            return new Vec3(testPos.getX() + 0.5D, testPos.getY(), testPos.getZ() + 0.5D);
        }
        return null;
    }

    private static Vec3 findVerticalFallback(ServerLevel level, Entity entity, LivingEntity anchor, Vec3 basePos, Vec3 facing) {
        BlockPos baseBlock = BlockPos.containing(basePos.x, basePos.y, basePos.z);
        for (int dy = 4; dy >= -4; dy--) {
            BlockPos testPos = baseBlock.offset(0, dy, 0);
            Vec3 candidate = new Vec3(testPos.getX() + 0.5D, testPos.getY(), testPos.getZ() + 0.5D);
            if (validateCandidate(level, entity, anchor, candidate, facing, true) != null) {
                return candidate;
            }
        }
        return null;
    }

    private static Vec3 findPointCandidate(ServerLevel level, Entity entity, Vec3 basePos) {
        BlockPos baseBlock = BlockPos.containing(basePos.x, basePos.y, basePos.z);
        for (int[] offset : POINT_OFFSETS) {
            for (int dy = 2; dy >= -3; dy--) {
                BlockPos testPos = baseBlock.offset(offset[0], dy, offset[1]);
                if (!canStandAt(level, entity, testPos)) {
                    continue;
                }
                return new Vec3(testPos.getX() + 0.5D, testPos.getY(), testPos.getZ() + 0.5D);
            }
        }
        return null;
    }

    private static boolean canStandAt(Level level, Entity entity, BlockPos pos) {
        if (level.isOutsideBuildHeight(pos) || level.isOutsideBuildHeight(pos.above())) {
            return false;
        }
        if (!level.getBlockState(pos.below()).entityCanStandOn(level, pos.below(), entity)) {
            return false;
        }
        if (!level.getBlockState(pos).canBeReplaced() || !level.getBlockState(pos.above()).canBeReplaced()) {
            return false;
        }

        double x = pos.getX() + 0.5D;
        double y = pos.getY();
        double z = pos.getZ() + 0.5D;
        return level.noCollision(entity, entity.getBoundingBox().move(
                x - entity.getX(),
                y - entity.getY(),
                z - entity.getZ()
        ));
    }
}

File: src/main/java/org/trp/shincolle/utility/ShipTeleportHelper.kt
package org.trp.shincolle.utility

import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.TamableAnimal
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.phys.Vec3
import org.trp.shincolle.item.DebugInspectorItem

object ShipTeleportHelper {
    private const val MIN_PLAYER_DISTANCE_SQ = 9.0
    private const val TARGET_REACH_DISTANCE_SQ = 144.0
    private val PREFERRED_OFFSETS = arrayOf<IntArray>(
        intArrayOf(-4, 0), intArrayOf(-4, -2), intArrayOf(-4, 2),
        intArrayOf(-5, 0), intArrayOf(-5, -3), intArrayOf(-5, 3),
        intArrayOf(-3, -3), intArrayOf(-3, 3), intArrayOf(-6, -1), intArrayOf(-6, 1),
        intArrayOf(-2, -4), intArrayOf(-2, 4)
    )
    private val FALLBACK_OFFSETS = arrayOf<IntArray>(
        intArrayOf(-7, 0), intArrayOf(-7, -2), intArrayOf(-7, 2),
        intArrayOf(-5, -5), intArrayOf(-5, 5), intArrayOf(-1, -5), intArrayOf(-1, 5),
        intArrayOf(0, -6), intArrayOf(0, 6), intArrayOf(-8, 0)
    )
    private val POINT_OFFSETS = arrayOf<IntArray>(
        intArrayOf(0, 0), intArrayOf(1, 0), intArrayOf(-1, 0), intArrayOf(0, 1), intArrayOf(0, -1),
        intArrayOf(2, 0), intArrayOf(-2, 0), intArrayOf(0, 2), intArrayOf(0, -2),
        intArrayOf(2, 1), intArrayOf(2, -1), intArrayOf(-2, 1), intArrayOf(-2, -1),
        intArrayOf(1, 2), intArrayOf(-1, 2), intArrayOf(1, -2), intArrayOf(-1, -2),
        intArrayOf(3, 0), intArrayOf(-3, 0), intArrayOf(0, 3), intArrayOf(0, -3)
    )

    @JvmStatic
    fun teleportNearLiving(entity: Entity, anchor: LivingEntity, verticalOffset: Double): Boolean {
        if (entity.level() !is ServerLevel) {
            return false
        }

        val basePos = Vec3(anchor.getX(), anchor.getY() + verticalOffset, anchor.getZ())
        val facing = anchor.getLookAngle()
        var horizontalFacing = Vec3(facing.x, 0.0, facing.z)
        if (horizontalFacing.lengthSqr() < 1.0E-4) {
            horizontalFacing = Vec3(0.0, 0.0, 1.0)
        } else {
            horizontalFacing = horizontalFacing.normalize()
        }
        val right = Vec3(-horizontalFacing.z, 0.0, horizontalFacing.x)

        var candidate =
            findCandidate(serverLevel, entity, anchor, basePos, horizontalFacing, right, PREFERRED_OFFSETS, true)
        if (candidate == null) {
            candidate =
                findCandidate(serverLevel, entity, anchor, basePos, horizontalFacing, right, FALLBACK_OFFSETS, true)
        }
        if (candidate == null) {
            candidate = findVerticalFallback(serverLevel, entity, anchor, basePos, horizontalFacing)
        }
        if (candidate == null) {
            return false
        }

        entity.teleportTo(candidate.x, candidate.y, candidate.z)
        notifyOwnerIfDebugging(entity)
        return true
    }

    @JvmStatic
    fun teleportNearPoint(entity: Entity, anchor: Vec3, verticalOffset: Double): Boolean {
        if (entity.level() !is ServerLevel) {
            return false
        }

        val candidate = findPointCandidate(serverLevel, entity, anchor.add(0.0, verticalOffset, 0.0))
        if (candidate == null) {
            return false
        }

        entity.teleportTo(candidate.x, candidate.y, candidate.z)
        notifyOwnerIfDebugging(entity)
        return true
    }

    private fun findCandidate(
        level: ServerLevel,
        entity: Entity,
        anchor: LivingEntity,
        basePos: Vec3,
        back: Vec3,
        right: Vec3,
        offsets: Array<IntArray>,
        rejectFront: Boolean
    ): Vec3? {
        for (offset in offsets) {
            val candidate = basePos
                .add(back.scale(offset[0].toDouble()))
                .add(right.scale(offset[1].toDouble()))
            val safePos = validateCandidate(level, entity, anchor, candidate, back, rejectFront)
            if (safePos != null) {
                return safePos
            }
        }
        return null
    }

    private fun validateCandidate(
        level: ServerLevel,
        entity: Entity,
        anchor: LivingEntity,
        candidate: Vec3,
        facing: Vec3,
        rejectFront: Boolean
    ): Vec3? {
        val dx = candidate.x - anchor.getX()
        val dz = candidate.z - anchor.getZ()
        val horizontalDistSq = dx * dx + dz * dz
        if (horizontalDistSq < MIN_PLAYER_DISTANCE_SQ) {
            return null
        }

        if (rejectFront) {
            val dot = dx * facing.x + dz * facing.z
            if (dot > 0.0) {
                return null
            }
        }

        if (anchor.distanceToSqr(candidate) > TARGET_REACH_DISTANCE_SQ) {
            return null
        }

        val baseBlock = BlockPos.containing(candidate.x, candidate.y, candidate.z)
        for (dy in 2 downTo -2) {
            val testPos = baseBlock.offset(0, dy, 0)
            if (!canStandAt(level, entity, testPos)) {
                continue
            }
            return Vec3(testPos.getX() + 0.5, testPos.getY().toDouble(), testPos.getZ() + 0.5)
        }
        return null
    }

    private fun findVerticalFallback(
        level: ServerLevel,
        entity: Entity,
        anchor: LivingEntity?,
        basePos: Vec3,
        facing: Vec3?
    ): Vec3? {
        val baseBlock = BlockPos.containing(basePos.x, basePos.y, basePos.z)
        for (dy in 4 downTo -4) {
            val testPos = baseBlock.offset(0, dy, 0)
            if (!canStandAt(level, entity, testPos)) {
                continue
            }
            return Vec3(testPos.getX() + 0.5, testPos.getY().toDouble(), testPos.getZ() + 0.5)
        }
        return null
    }

    private fun findPointCandidate(level: ServerLevel, entity: Entity, basePos: Vec3): Vec3? {
        val baseBlock = BlockPos.containing(basePos.x, basePos.y, basePos.z)
        for (offset in POINT_OFFSETS) {
            for (dy in 2 downTo -3) {
                val testPos = baseBlock.offset(offset[0], dy, offset[1])
                if (!canStandAt(level, entity, testPos)) {
                    continue
                }
                return Vec3(testPos.getX() + 0.5, testPos.getY().toDouble(), testPos.getZ() + 0.5)
            }
        }
        return null
    }

    private fun canStandAt(level: Level, entity: Entity, pos: BlockPos): Boolean {
        if (level.isOutsideBuildHeight(pos) || level.isOutsideBuildHeight(pos.above())) {
            return false
        }
        if (!level.getBlockState(pos.below()).entityCanStandOn(level, pos.below(), entity)) {
            return false
        }
        if (!level.getBlockState(pos).canBeReplaced() || !level.getBlockState(pos.above()).canBeReplaced()) {
            return false
        }

        val x = pos.getX() + 0.5
        val y = pos.getY().toDouble()
        val z = pos.getZ() + 0.5
        return level.noCollision(
            entity, entity.getBoundingBox().move(
                x - entity.getX(),
                y - entity.getY(),
                z - entity.getZ()
            )
        )
    }

    private fun notifyOwnerIfDebugging(entity: Entity?) {
        if (entity !is TamableAnimal) return
        val ownerId = entity.getOwnerUUID()
        if (ownerId == null) return
        if (entity.level() !is ServerLevel) return
        val player: ServerPlayer? = sl.getServer().getPlayerList().getPlayer(ownerId)
        if (player == null) return
        if (!hasDebugInspectorEquipped(player)) return
        val pos = String.format("(%.2f, %.2f, %.2f)", entity.getX(), entity.getY(), entity.getZ())
        player.sendSystemMessage(
            Component.literal(
                "[DebugInspector] Ship " + entity.getUUID().toString().substring(0, 8) + "... teleported to " + pos
            ), false
        )
    }

    private fun hasDebugInspectorEquipped(player: Player): Boolean {
        if (player.getOffhandItem().getItem() is DebugInspectorItem) return true
        for (stack in player.getArmorSlots()) {
            if (stack.getItem() is DebugInspectorItem) return true
        }
        return false
    }
}
